File: spring-ceph-client/src/test/java/com/example/httpclient/jdk/JdkRestClientTests.java
package com.example.httpclient.jdk;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import com.example.httpclient.restclient.jdk.RestClient;
import com.example.httpclient.restclient.jdk.RestClient.HttpMethod;
import tools.jackson.databind.json.JsonMapper;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@SpringBootTest
public class JdkRestClientTests {

    private RestClient client = RestClient.INSTANCE;

    @Autowired
    private JsonMapper jsonMapper;

    @Test
    public void head() {
        var url = "http://localhost:8181/api/pagenation/servers/count";
        var response = client.send(HttpMethod.HEAD, url, null, "{}");
        if (response.is2xxSuccessful && response.body.length > 0) {
            log.info("Successfully received response with status body=\n{}", new String(response.body));
        } else {
            log.error("Failed to receive successful response.");
        }
    }

    @Test
    public void count() {
        var url = "http://localhost:8181/api/pagenation/servers/count";
        var response = client.send(HttpMethod.POST, url, null, "{}");
        if (response.is2xxSuccessful && response.body.length > 0) {
            log.info("Successfully received response with status body=\n{}", new String(response.body));
        } else {
            log.error("Failed to receive successful response.");
        }
    }

    @Test
    public void pagenation() {
        var countUrl = "http://localhost:8181/api/pagenation/servers/count";
        var serverUrl = "http://localhost:8181/api/pagenation/servers";

        var response = client.send(HttpMethod.POST, countUrl, null, "{}");
        log.info("response=\n{}", new String(response.body));

        int totalCount = 0;
        try {
            totalCount = jsonMapper.readTree(response.body).at("/body/totalCount").asInt();
            log.info("totalCount={}", totalCount);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            return;
        }

        String requestBody = """
        {
            "pageNumber": %d,
            "pageSize": %d
        }
        """;

        var results = client.postAllPages(serverUrl, null, requestBody, totalCount, totalCount/5, 3000);
        long count = results.values().stream().filter(b -> b.length > 0).count();

        log.info("received {} pages of servers", count);
        if (count <= 0) {
            log.error("can NOT receive any server list");
            return;
        }

        results.forEach((k, v) -> {
            if (v.length > 0) {
                log.info("page={}, body=\n{}", k, new String(v));
            } else {
                log.warn("page={}, body is empty", k);
            }
        });

        // Do something with the results, e.g., parse JSON and process server lists
        // results.values().stream().filter(b -> b.length > 0).forEach(b -> {
        //     try {
        //         log.info("body=\n{}", jsonMapper.writerWithDefaultPrettyPrinter().writeValueAsString(jsonMapper.readTree(b)));
        //     } catch (Exception e) {
        //         log.error(e.getMessage(), e);
        //     }
        // });
    }
}

File: spring-httpclient-kotlin/src/test/kotlin/com/example/httpclient/jdk/JdkRestClientTests.kt
package com.example.httpclient.jdk

import com.example.httpclient.restclient.jdk.RestClient
import com.example.httpclient.restclient.jdk.RestClient.HttpMethod
import com.fasterxml.jackson.databind.json.JsonMapper
import io.github.oshai.kotlinlogging.KotlinLogging
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class JdkRestClientTests {
    private val log = KotlinLogging.logger {}
    private val client = RestClient

    @Autowired
    private lateinit var jsonMapper: JsonMapper

    @Test
    fun head() {
        val url = "http://localhost:8181/api/pagenation/servers/count"
        val response = client.send(HttpMethod.HEAD, url, null, "{}")
        if (response.is2xxSuccessful && response.body.isNotEmpty()) {
            log.info { "Successfully received response with status body=\n${String(response.body)}" }
        } else {
            log.error { "Failed to receive successful response." }
        }
    }

    @Test
    fun count() {
        val url = "http://localhost:8181/api/pagenation/servers/count"
        val response = client.send(HttpMethod.POST, url, null, "{}")
        if (response.is2xxSuccessful && response.body.isNotEmpty()) {
            log.info { "Successfully received response with status body=\n${String(response.body)}" }
        } else {
            log.error { "Failed to receive successful response." }
        }
    }

    @Test
    fun pagenation() {
        val countUrl = "http://localhost:8181/api/pagenation/servers/count"
        val serverUrl = "http://localhost:8181/api/pagenation/servers"

        val response = client.send(HttpMethod.POST, countUrl, null, "{}")
        log.info { "response=\n${String(response.body)}" }

        val totalCount = try {
            val count = jsonMapper.readTree(response.body).at("/body/totalCount").asInt()
            log.info { "totalCount=$count" }
            count
        } catch (e: Exception) {
            log.error(e) { e.message ?: "failed to parse response" }
            return
        }

        val requestBody = """
        {
            "pageNumber": %d,
            "pageSize": %d
        }
        """.trimIndent()

        val results = client.postAllPages(serverUrl, null, requestBody, totalCount, totalCount / 5, 3000)
        val count = results.values.count { it.isNotEmpty() }

        log.info { "received $count pages of servers" }
        if (count <= 0) {
            log.error { "can NOT receive any server list" }
            return
        }

        results.forEach { (k, v) ->
            if (v.isNotEmpty()) {
                log.info { "page=$k, body=\n${String(v)}" }
            } else {
                log.warn { "page=$k, body is empty" }
            }
        }
    }
}
